ql/src/java/org/apache/hadoop/hive/ql/history/HiveScribeImpl.java

```diff
@@ -0,0 +1,331 @@
+      }
+      stats.setCurrentTimeStamp(sessionState.getQueryCurrentTimestamp().toString());
+      stats.setMapReduceStats(sessionState.getMapRedStats());
+    }
+  }
+
+  @Override
+  public void endQuery (String queryId) {
+    Long timeStamp = System.currentTimeMillis();
+    HiveHistory.QueryInfo queryInfo = queryInfoMap.get(queryId);
+    QueryStats stats = queryStatsMap.get(queryId);
+    stats.setQueryEnd(timeStamp);
+    addSessionInfo(queryId);
+    scribeQueryMetricEntry(queryId);
+    queryInfoMap.remove(queryId);
+    queryStatsMap.remove(queryId);
+  }
+
+  @Override
+  public void startTask(String queryId, Task<? extends Serializable> task, String taskName) {
+    String timeStamp = Long.toString(System.currentTimeMillis());
+    HiveHistory.TaskInfo taskInfo = new HiveHistory.TaskInfo();
+    taskInfo.hm.put(HiveHistory.Keys.QUERY_ID.name(), queryId);
+    taskInfo.hm.put(HiveHistory.Keys.TASK_ID.name(), task.getId());
+    taskInfo.hm.put(HiveHistory.Keys.TASK_NAME.name(), taskName);
+    String id = queryId + ":" + task.getId();
+    taskInfoMap.put(id, taskInfo);
+    QueryStats stats = queryStatsMap.get(queryId);
+    snapshotTaskProgress(RecordTypes.TaskStart, stats, taskInfo.hm, timeStamp);
+  }
+
+  @Override
+  public void endTask(String queryId, Task<? extends Serializable> task) {
+    String timeStamp = Long.toString(System.currentTimeMillis());
+    String id = queryId + ":" + task.getId();
+    HiveHistory.TaskInfo taskInfo = taskInfoMap.get(id);
+    if (taskInfo == null) {
+      return;
+    }
+
+    QueryStats stats = queryStatsMap.get(queryId);
+    snapshotTaskProgress(RecordTypes.TaskEnd, stats, taskInfo.hm, timeStamp);
+    taskInfoMap.remove(id);
+  }
+
+  @Override
+  public void progressTask(String queryId, Task<? extends Serializable> task) {
+    String timeStamp = Long.toString(System.currentTimeMillis());
+    String id = queryId + ":" + task.getId();
+    TaskInfo taskInfo = taskInfoMap.get(id);
+    if (taskInfo == null) {
+      return;
+    }
+
+    QueryStats stats = queryStatsMap.get(queryId);
+    snapshotTaskProgress(RecordTypes.TaskProgress, stats, taskInfo.hm, timeStamp);
+  }
+
+  public void snapshotTaskProgress(RecordTypes recordTypes, QueryStats stats, Map<String, String> taskStats, String timeStamp){
+    StringBuilder snapshot = new StringBuilder("");
+    snapshot.append(recordTypes.name());
+    for (Map.Entry<String, String> ent : taskStats.entrySet()) {
+      String key = ent.getKey();
+      String val = ent.getValue();
+      if(val != null) {
+        val = val.replace(System.getProperty("line.separator"), " ");
+      }
+      snapshot.append(" ");
+      snapshot.append(key + "=\"" + val + "\"");
+    }
+    insertTaskProgress(stats.getTaskProgress(), timeStamp, snapshot.toString());
+  }
+
+  private void insertTaskProgress(Map<String, String> taskProgressStats, String timeStamp, String taskProgress) {
+    for (Map.Entry<String, String> ent : taskProgressStats.entrySet()) {
```

Comment:
This is very inefficient. A better way to achieve the goal would probably be 
 - store latest `taskProgress` (may need to store one for each of active task)
 - compare the incoming taskProgress with the last `taskProgress`
   - if the same, skip it
   - otherwise, add it to the taskProgressStats

This actually makes me wonder if storing the taskProgress in a list would be a better idea.